### PR TITLE
add git commit signing

### LINKED ISSUE #126: Add option to enable commit signing
# Context

At the moment, if you use this action and have the `Require signed commits` setting enabled for a protected branch, the PR generated can't be merged.

# Example

[Crowdin](https://crowdin.com/), a popular localisation management tool, has a similar [action](https://github.com/crowdin/github-action) which opens a PR using a bot.

They allow you to pass a `gpg_private_key` as a config option, which will enable signed commits for the bot opening the PR. See options [here](https://github.com/crowdin/github-action#supported-options).

### PR DESCRIPTION
**Pull request by jacklesliewise**
# Context

Fixes #126 

# Changes

- Add a `githubUserName` input option
  - Default to `"github-actions[bot]"`
- Add a `githubUserEmail` input option
  - Default to `"github-actions[bot]@users.noreply.github.com"`
- Add a `GPG_PRIVATE_KEY` env option
  - This determines if keys should be signed or not
- Pass in `githubUserName` and `githubUserEmail` to `setupUser`
  - Can be done outwith signed commits, but these are needed to verify the signed commits if its enabled
- Add a `setupCommitSigning` if `GPG_PRIVATE_KEY` is present
  - Need to install `gpg`
  - use `STDIN` for `gpg --import` (utf-8 encoding with `Buffer.from` by default)
  - Need to do a `/bin/bash` command to get and grep the key id - `exec` doesn't support piping and multiple commands on its own
  - Set `git` config for `user.signingkey` and `commit.gpgsign`

# Considerations

- Input option naming (`githubUserName` vs `github_user_name`)
- Should the input options just be more `env` variables instead?]
- How do you feel about installing `gpg` here?

### CODE REVIEW

File: src/gitUtils.ts [hunk 1]
```diff
@@ -1,21 +1,43 @@
...
     "config",
     "--global",
     "user.email",
-    `"github-actions[bot]@users.noreply.github.com"`,
+    githubUserEmail,
   ]);
 };
 
+export const setupCommitSigning = async (gpgPrivateKey: string) => {
+  await exec("apk", [
```
Comment (Andarist) [inline src/gitUtils.ts:20]: is apk even available by default on ubuntu? why the Crowdin action doesn't have to do this installation at all?

File: src/gitUtils.ts [hunk 2]
```diff
@@ -1,21 +1,43 @@
...
+    "add",
+    "--no-cache",
+    "gnupg",
+  ]);
+  await exec("gpg", ["--import"], { input: Buffer.from(gpgPrivateKey) });
+  const { stdout: keyId } = await execWithOutput(`/bin/bash -c "gpg --list-secret-keys --with-colons | grep '^sec:' | cut -d ':' -f 5"`);
```
Comment (Andarist) [inline src/gitUtils.ts:26]: couldnt this be decoded from the importing's command output? something like this should display the information about what is being imported
```
gpg --with-colons --import-options show-only --import
```

the problem with listing all keys is that potentially the returned list could potentially contain many entries and we could grab some unrelated `keyId` here

File: src/index.ts [hunk 3]
```diff
@@ -14,8 +14,19 @@ const getOptionalInput = (name: string) => core.getInput(name) || undefined;
     return;
   }
 
+  let githubUserName = getOptionalInput('githubUserName') || `"github-actions[bot]"`;
+  let githubUserEmail = getOptionalInput('githubUserEmail') || `"github-actions[bot]@users.noreply.github.com"`;
```
Comment (Andarist) [inline src/index.ts:18]: those are not related to github, so I think this would be more appropriate:
```suggestion
  let gitUserName = getOptionalInput('gitUserName') || `"github-actions[bot]"`;
  let gitUserEmail = getOptionalInput('gitUserEmail') || `"github-actions[bot]@users.noreply.github.com"`;
```

File: src/index.ts [hunk 4]
```diff
@@ -14,8 +14,19 @@ const getOptionalInput = (name: string) => core.getInput(name) || undefined;
...
+
   console.log("setting git user");
-  await gitUtils.setupUser();
+ 
+  await gitUtils.setupUser(githubUserName, githubUserEmail);
```
Comment (Andarist) [inline src/index.ts:22]: nit: we could just group those nicely, like this:
```suggestion
  await gitUtils.setupUser({ name: gitUserName, email: gitUserEmail });
```

File: src/gitUtils.ts (hunk 1 above)
Comment (jacklesliewise) [reply src/gitUtils.ts:20]: The crowdin action does this installation in the [`Dockerfile`](https://github.com/crowdin/github-action/blob/master/Dockerfile)

What would the alternative be? Have a prerequisite that if you want commit signing the image you're using should have it installed already?

File: src/gitUtils.ts (hunk 2 above)
Comment (jacklesliewise) [reply src/gitUtils.ts:26]: Yeah true, although the current implementation seems to be sufficient for the crowdin action, especially if consumers are just passing the one private key.

Happy to explore your solution though if this is a blocker.

File: src/index.ts (hunk 3 above)
Comment (jacklesliewise) [reply src/index.ts:18]: Makes sense, will update these along with the README.

File: src/gitUtils.ts (hunk 1 above)
Comment (Andarist) [reply src/gitUtils.ts:20]: > What would the alternative be? Have a prerequisite that if you want commit signing the image you're using should have it installed already?

Yes, this sounds reasonable to me - I agree that it would be better if we could handle this automatically, but installing tools seems out of the scope of this simple node.js action. Note that most of our users use the default ubuntu images but it's technically possible that this action gets executed in different contexts too - so I'm afraid that we would always fail to recognize a particular environment and thus we'd fail to install this correctly on the user's behalf.

File: src/gitUtils.ts (hunk 2 above)
Comment (Andarist) [reply src/gitUtils.ts:26]: It's not a blocker per se but if this solution could be used then I would prefer using it - cause it seems to be more explicit about the whole flow. It's less dependent on other factors and the preexisting environment.

File: src/gitUtils.ts (hunk 1 above)
Comment (jacklesliewise) [reply src/gitUtils.ts]: Agreed, I'll remove this line then and add something to the `README`

File: src/gitUtils.ts (hunk 2 above)
Comment (jacklesliewise) [reply src/gitUtils.ts]: Yeah that makes sense, I won't pick this up for a few days now but I've addressed all the other comments, if it's sufficient as is then feel free to merge but if not I'll come back to this

**Comment by mitchellhamilton** [comment]
Could we just have an option to disable setting the git email and name and then you can setup the signing before the Changesets action runs? It seems strange to have this built-in to the Changesets action.

**Comment by jacklesliewise** [comment]
@mitchellhamilton yeah that's valid, I'm happy to open up a new PR with that approach if you prefer.

The current implementation is based on the Crowdin action, so it's up to you whether you'd like to align with that or if you'd rather have the action assume the git user is set up.

**Comment by jacklesliewise** [comment]
@mitchellhamilton here's another [PR](https://github.com/changesets/action/pull/131) with the solution you suggested

Let me know which approach you both prefer

**Comment by jacklesliewise** [comment]
Closing in favour of #131 131
